src/App.tsx

```diff
@@ -1,24 +1,241 @@
+      setErrorMessage('Unable to delete a todo');
+      warningTimer(setHasError, false, 3000);
+    }
+  };
+
+  const allCompleted = todos.filter(todo => todo.completed === true);
+
+  const isAllCompleted = allCompleted.length === todos.length;
+
+  const clearCompleted = () => {
+    allCompleted.forEach(todo => deleteTodo(USER_ID, todo.id));
+    setTodos(prevTodos => prevTodos.filter(todo => todo.completed === false));
+  };
+
+  const onToogleTodo = async (todoTogle: Todo) => {
+    try {
+      const todoChangeStatus = await toogleTodo(USER_ID, todoTogle.id, !todoTogle.completed);
+
+      setTodos(currentTodos => currentTodos.map(todo => {
+        return todo.id === todoChangeStatus.id
+          ? todoChangeStatus
+          : todo;
+      }));
+    } catch (error) {
+      setHasError(true);
+      setErrorMessage('Unable to change completed');
+      warningTimer(setHasError, false, 3000);
+    }
+  };
+
+  const toogleAllTodo = () => {
+    if (isAllCompleted) {
+      todos.map(async (todoToogle) => {
+        try {
+          const todoChangeStatus = await toogleTodo(USER_ID, todoToogle.id, false);
+
+          setTodos(currentTodos => currentTodos.map(todo => {
+            return todo.id === todoChangeStatus.id
+              ? todoChangeStatus
+              : todo;
+          }));
+        } catch (error) {
+          setHasError(true);
+          setErrorMessage('Unable to change completed');
+          warningTimer(setHasError, false, 3000);
+        }
+      });
+    } else {
+      todos.map(async todoToogle => {
+        try {
+          const todoChangeStatus = await toogleTodo(USER_ID, todoToogle.id, true);
+
+          setTodos(currentTodos => currentTodos.map(todo => {
+            return todo.id === todoChangeStatus.id
+              ? todoChangeStatus
+              : todo;
+          }));
+        } catch (error) {
+          setHasError(true);
+          setErrorMessage('Unable to change completed');
+          warningTimer(setHasError, false, 3000);
+        }
+      });
+    }
+  };
+
+  const handleUpdateTodo = useCallback(async (todoToUpdate: Todo) => {
+    try {
+      setTodosInProcessed(currentTodos => [...currentTodos, todoToUpdate]);
+
+      const updatedTodo = await updateTodo(USER_ID, todoToUpdate);
+
+      setTodos(currentTodos => currentTodos.map(todo => {
+        return todo.id === updatedTodo.id
+          ? updatedTodo
+          : todo;
+      }));
+    } catch (error) {
+      setHasError(true);
+      setErrorMessage('Unable to delete a todo');
+      warningTimer(setHasError, false, 3000);
+    }
+  }, []);
+
+  const visibleTodos = useMemo(() => {
+    return todos.filter(todo => {
+      const completedTodo = todo.completed;
+
+      switch (selectFilter) {
+        case 'all':
+          return todo;
+        case 'active':
+          return !completedTodo;
+        case 'completed':
+          return completedTodo;
+        default:
+          return false;
+      }
+    });
```

Comment:
make separated filters in each case, now it's a wasting of performance for all & default cases